### PR TITLE
v0.02 complete

### PR DESCRIPTION
**Pull request by JodyLS27**
I have completed phase two which includes correct collision with the window, AABB Collision with the target, a score counter for the player, and improved movement of the player object.

### CODE REVIEW

File: src/player.cpp [hunk 1]
```diff
@@ -0,0 +1,117 @@
+#include <player.hpp>
+
+
+player::Player::Player(int score, int speed)
+	: m_score(score), m_speed(speed), m_width(25), m_height(25)
```
Comment (swagween) [inline src/player.cpp]: you could set `m_width` and `m_height` using optional constructor arguments instead of hard-coding it

File: src/player.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+void player::Player::set_shape(sf::Color color)
+{
+	m_shape.setSize({ static_cast<float>(m_width), static_cast<float>(m_height) });
+	m_shape.setOrigin({ m_width * 0.5f, m_height * 0.5f });
+	m_shape.setPosition(m_position);
+	m_shape.setFillColor(color);
+}
+
+void player::Player::set_score()
```
Comment (swagween) [inline src/player.cpp]: I would rename this function to `increment_score()`

File: src/player.cpp [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+}
+
+// Set the position the player is going to be and then check if it hits the 
+// window and adjust the positioning.
+sf::Vector2f player::Player::window_collision(const sf::Vector2u window_size,
+	float delta_time)
+{
+	sf::Vector2f ret_position = m_position + (m_direction * delta_time) * m_speed;
+
+	if (ret_position.x < (m_width * 0.5f)) { ret_position.x = (m_width * 0.5f); }
```
Comment (swagween) [inline src/player.cpp]: consider writing a `get_center()` method so you don't have to keep writing `m_width * 0.5f`

File: src/player.hpp [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#pragma once
+#include <SFML/Graphics.hpp>
+
+namespace player
+{
+	class Player
```
Comment (swagween) [inline src/player.hpp]: `namespace` and its highest level code block (in this case, `class Player {}`) should have the same indentation

File: src/target.cpp [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+{
+	return m_shape;
+}
+
+sf::Vector2f target::Target::generate_random_position(const sf::Vector2u window_size)
+{
+	sf::Vector2f position{ static_cast<float>(random_val(0, window_size.x)), static_cast<float>(random_val(0, window_size.y)) };
+
+	// Bounds checking.
+	if (position.x < m_radius) { position.x = m_radius; }
```
Comment (swagween) [inline src/target.cpp]: this is sorta duplicated code; in the future, you could consider writing some kind of `collider` component that has a method with this code, and then give both `target` and `player` a member `Collider` and check window collision against that.

File: src/main.cpp [hunk 6]
```diff
@@ -1,6 +1,108 @@
...
+	window.draw(text);
+	window.display();
+}
 
 int main()
 {
-    return 0;
-}
\ No newline at end of file
+	player::Player player(0, 100.0f);
```
Comment (swagween) [inline src/main.cpp]: I thought this should to be initialized with `{}` but I'm not really sure if it makes a difference

File: src/main.cpp [hunk 7]
```diff
@@ -1,6 +1,108 @@
...
+		{
+			switch (event.type)
+			{
+			case sf::Event::Closed:
+				window.close();
+				break;
+			}
+		}
+
+		logic(player, target, event, window_size, delta_time);
```
Comment (swagween) [inline src/main.cpp]: I think you want to handle events separately from logic. `Player::event_handler()` should be called whilst the window is polling the event queue. Logic can be applied after all events are processed. I'm not sure about the consequences of doing it your way, and it's possible that it's totally fine, but it seems like it could lead to some unexpected behavior.

File: src/main.cpp [hunk 8]
```diff
@@ -1,6 +1,108 @@
...
+	if (!font.loadFromFile("C:/Users/jodys/OneDrive/Documents/Fonts/Vogue.ttf"))
+	{
+		std::cout << "Error: Cannot load file." << std::endl;
+	}
+
+	using clock = std::chrono::steady_clock;
+	using mil = std::chrono::milliseconds;
+	using flt = std::chrono::duration<float>;
+
+	auto end_clock = std::chrono::time_point_cast<mil>(clock::now());
```
Comment (karnkaul) [inline src/main.cpp:164]: Why is this needed?

File: src/main.cpp [hunk 9]
```diff
@@ -1,6 +1,108 @@
...
+
+	init(window_size, player, target);
+
+	while (window.isOpen())
+	{
+		sf::Event event;
+
+		auto current_clock = std::chrono::time_point_cast<mil>(clock::now());
```
Comment (karnkaul) [inline src/main.cpp:180]: The cast is unnecessary, you can just subtract two time points implicitly into a float duration.

File: src/main.cpp (hunk 7 above)
Comment (karnkaul) [reply src/main.cpp]: +1
Store the state changes in the event loop, then respond to those changes later in the tick / update loop.

File: src/player.cpp (hunk 1 above)
Comment (karnkaul) [reply src/player.cpp]: Even better: set them in the class definition itself

File: src/player.cpp [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	m_position = position;
+}
+
+void player::Player::set_direction()
+{
+	m_direction = { 0.0f, 0.0f };
+	if (m_move_up) { m_direction = { 0.0f, -1.0f }; }
+	if (m_move_left) { m_direction = { -1.0f, 0.0f }; }
+	if (m_move_down) { m_direction = { 0.0f, 1.0f }; }
+	if (m_move_right) { m_direction = { 1.0f , 0.0f }; }
```
Comment (karnkaul) [inline src/player.cpp]: What about diagonals?

File: src/player.cpp [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+{
+	m_shape.setSize({ static_cast<float>(m_width), static_cast<float>(m_height) });
+	m_shape.setOrigin({ m_width * 0.5f, m_height * 0.5f });
+	m_shape.setPosition(m_position);
+	m_shape.setFillColor(color);
+}
+
+void player::Player::set_score()
+{
+	m_score++;
```
Comment (karnkaul) [inline src/player.cpp]: Use pre increment by default

File: src/player.cpp [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+{
+	return m_width;
+}
+
+int player::Player::get_height() const
+{
+	return m_height;
+}
+
+void player::Player::event_handler(sf::Event event)
```
Comment (karnkaul) [inline src/player.cpp]: Don't copy events

File: src/player.cpp [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+		if (event.key.code == sf::Keyboard::S || event.key.code == sf::Keyboard::Down) { m_move_down = true; }
+		if (event.key.code == sf::Keyboard::D || event.key.code == sf::Keyboard::Right) { m_move_right = true; }
+	}
+
+	if (event.type == sf::Event::KeyReleased)
+	{
+		if (event.key.code == sf::Keyboard::W || event.key.code == sf::Keyboard::Up) { m_move_up = false; }
+		if (event.key.code == sf::Keyboard::A || event.key.code == sf::Keyboard::Left) { m_move_left = false; }
+		if (event.key.code == sf::Keyboard::S || event.key.code == sf::Keyboard::Down) { m_move_down = false; }
+		if (event.key.code == sf::Keyboard::D || event.key.code == sf::Keyboard::Right) { m_move_right = false; }
```
Comment (karnkaul) [inline src/player.cpp]: Improve this mess of `if`s 😄

File: src/player.cpp [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+	}
+}
+
+void player::Player::update_position(sf::Vector2f position)
+{
+	m_position = position;
```
Comment (karnkaul) [inline src/player.cpp]: Why is `m_position` a separate variable when `m_shape.getPosition()` is sufficient?

File: src/player.cpp [hunk 15]
```diff
@@ -0,0 +1,117 @@
...
+	m_shape.setPosition(position);
+}
+
+void player::Player::update_position(float delta_time)
```
Comment (karnkaul) [inline src/player.cpp]: Take `chrono::duration<float>` as `delta_time`; `float` doesn't indicate what the unit is.

File: src/player.cpp [hunk 16]
```diff
@@ -0,0 +1,117 @@
...
+// window and adjust the positioning.
+sf::Vector2f player::Player::window_collision(const sf::Vector2u window_size,
+	float delta_time)
+{
+	sf::Vector2f ret_position = m_position + (m_direction * delta_time) * m_speed;
+
+	if (ret_position.x < (m_width * 0.5f)) { ret_position.x = (m_width * 0.5f); }
+	if (ret_position.x > (window_size.x - (m_width * 0.5f))) { ret_position.x = (window_size.x - (m_width * 0.5f)); }
+	if (ret_position.y < (m_height * 0.5f)) { ret_position.y = (m_height * 0.5f); }
+	if (ret_position.y > (window_size.y - (m_height * 0.5f))) { ret_position.y = (window_size.y - (m_height * 0.5f)); }
```
Comment (karnkaul) [inline src/player.cpp]: Again lots of `if`s

File: src/player.hpp [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+	{
+	private:
```
Comment (karnkaul) [inline src/player.hpp]: You'd generally want to put `public` stuff first, as its part of the class' API, and `private` at the bottom.

File: src/target.cpp [hunk 18]
```diff
@@ -0,0 +1,87 @@
+#include <target.hpp>
+
+target::Target::Target(float radius)
+	: m_radius(radius)
+{}
+
+int ::target::Target::random_val(int lo, int hi)
+{
+	auto engine = std::default_random_engine{ std::random_device{}() };
```
Comment (karnkaul) [inline src/target.cpp]: This is incorrect, you're creating a new `engine` on every call. You want to use the same engine that's initialized with a seed _once_ throughout the entire program.

File: src/target.cpp [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+	auto distribution = std::uniform_int_distribution<int>{ lo, hi };
+	return distribution(engine);
+}
+
+
+sf::Vector2f target::Target::get_position()
```
Comment (karnkaul) [inline src/target.cpp]: Missing `const`

File: src/target.cpp [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+sf::Vector2f target::Target::generate_random_position(const sf::Vector2u window_size)
+{
+	sf::Vector2f position{ static_cast<float>(random_val(0, window_size.x)), static_cast<float>(random_val(0, window_size.y)) };
+
+	// Bounds checking.
+	if (position.x < m_radius) { position.x = m_radius; }
+	if (position.x > window_size.x - (m_radius * 1.5f)) { position.x = window_size.x - (m_radius * 1.5f); }
+	if (position.y < m_radius) { position.y = m_radius; }
+	if (position.y > window_size.y - (m_radius * 1.5f)) { position.y = window_size.y - (m_radius * 1.5f); }
```
Comment (karnkaul) [inline src/target.cpp]: Lots of `if`s

File: src/target.cpp [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+
+	return position;
+}
+
+bool target::Target::collision_hit(const sf::Vector2f player_position, const int player_width, const int player_height)
+{
+	/// <summary>
```
Comment (karnkaul) [inline src/target.cpp]: Doc comments go above the function / class, not inside

File: src/main.cpp [hunk 22]
```diff
@@ -1,6 +1,108 @@
...
+	player::Player player(0, 100.0f);
+	target::Target target(15.0f);
+
+	const sf::Vector2u window_size = sf::Vector2u{ 800, 600 };
+	sf::RenderWindow window(sf::VideoMode(window_size.x, window_size.y), "Move To Target");
+
+	sf::Font font;
+	sf::Text text;
+
+	if (!font.loadFromFile("C:/Users/jodys/OneDrive/Documents/Fonts/Vogue.ttf"))
```
Comment (karnkaul) [inline src/main.cpp]: Pretty sure this file isn't going to exist anywhere except your dev machine 😄

File: src/player.hpp (hunk 4 above)
Comment (JodyLS27) [reply src/player.hpp]: Hey there.
Yea these are the indentations from Visual Studio... I need to look into a different styling as this is what it gives me even if I put it with the same indentations.

Comment (JodyLS27) [reply src/player.hpp]: @karnkaul Should I be using Clang formatting ?

File: src/target.cpp (hunk 5 above)
Comment (JodyLS27) [reply src/target.cpp]: Hey :)

So I am going to be implementing a collision class that will deal with this so I can just pass objects in and check their collision state. It will help with redundancy and clean it up a bit 😄

File: src/main.cpp (hunk 6 above)
Comment (JodyLS27) [reply src/main.cpp]: Yea I did some looking and your right, Will adjust thank you 👍

File: src/main.cpp (hunk 7 above)
Comment (JodyLS27) [reply src/main.cpp]: Hey 😃 
So If I place the player.event_handler(event) in the while loop for the pollEvent(), I get unexpected behavior where the player object starters each time a different key is pressed plus there is a strong lag due to the loop.
So placing it outside helped smooth out the movement.

Not sure if I should go back to default and just have a case key switch event in the main window ?

Comment (swagween) [reply src/main.cpp]: the reason for that is probably this part of your `set_direction()` function.
```cpp
m_direction = { 0.0f, 0.0f };
if (m_move_up) { m_direction = { 0.0f, -1.0f }; }
if (m_move_left) { m_direction = { -1.0f, 0.0f }; }
if (m_move_down) { m_direction = { 0.0f, 1.0f }; }
if (m_move_right) { m_direction = { 1.0f , 0.0f }; }
```
Think about what will happen here if the player presses and holds the `right` key, then also presses the `up` key. The way you have it here, there is a different priority for directions simply because of how they're ordered in the code. `right` beats everything because it's checked last, and `m_direction` is overwritten regardless of the other movement states.

Either way, you should switch to event handling inside the loop and troubleshoot the bugs from there. It's weird that you get a lag like you said, but I highly doubt it's from event polling because that happens pretty much instantly.

Comment (JodyLS27) [reply src/main.cpp]: Cool, yea I will look into this further, Thank you.

File: src/main.cpp (hunk 9 above)
Comment (JodyLS27) [reply src/main.cpp:180]: Hey Karn, 
I did this casting as to not have the extra floating points in the time, but if its not needed to cast, I will look into remove it.

File: src/main.cpp (hunk 8 above)
Comment (JodyLS27) [reply src/main.cpp:164]: Hey Karn, 
Are you referring to the float variable?

File: src/main.cpp (hunk 9 above)
Comment (karnkaul) [reply src/main.cpp:180]: Why do you not want extra floating points in time? That will make the simulation inaccurate.

File: src/player.cpp (hunk 13 above)
Comment (JodyLS27) [reply src/player.cpp]: Yup... On it :)

File: src/player.cpp (hunk 14 above)
Comment (JodyLS27) [reply src/player.cpp]: Yea... I could do that 🤔 Thanks Karn

File: src/main.cpp (hunk 8 above)
Comment (karnkaul) [reply src/main.cpp:164]: I see that you're assigning to it inside the main loop, that's cool then

File: src/main.cpp (hunk 9 above)
Comment (karnkaul) [reply src/main.cpp:180]: This is still truncating decimals

File: src/main.cpp [hunk 23]
```diff
@@ -1,6 +1,183 @@
...
+		auto current_clock = std::chrono::time_point_cast<mil>(clock::now());
+		auto delta_time = current_clock - end_clock;
+		end_clock = current_clock;
+
+		event_handling(window, event, player);
+
+		logic(player, target, window_size, delta_time);
+
+		draw_text(font, text, player.get_score());
+		render(window, player.get_shape(), target.get_shape(), text);
```
Comment (karnkaul) [inline src/main.cpp:189]: Recommend refactoring this logic to be in a class that stores font, text, player, target, etc as member variables, and takes in delta time in its update/tick function.

File: src/main.cpp [hunk 24]
```diff
@@ -1,6 +1,183 @@
...
+			window.close();
+			break;
+
+			// Character Movement
+		case sf::Event::KeyPressed:
+			switch (event.key.code)
+			{
+			case sf::Keyboard::Up:
+			case sf::Keyboard::W:
+				if (player.can_change_direction("up"))
```
Comment (karnkaul) [inline src/main.cpp:37]: These string comparisons are unnecessary and they are also happening every frame. Use an enum instead.

File: src/player.cpp [hunk 25]
```diff
@@ -0,0 +1,200 @@
...
+		return;
+	}
+
+	if (m_move_right)
+	{
+		m_direction = { 1.0f , 0.0f };
+		return;
+	}
+
+	m_direction = { 0.0f , 0.0f };
```
Comment (karnkaul) [inline src/player.cpp:96]: Why not just update `m_direction` earlier and then normalize it here?

File: src/player.cpp [hunk 26]
```diff
@@ -0,0 +1,200 @@
...
+
+	if (ret_position.y < (m_height * 0.5f))
+	{
+		ret_position.y = (m_height * 0.5f);
+	}
+
+	if (ret_position.y > (window_size.y - (m_height * 0.5f)))
+	{
+		ret_position.y = (window_size.y - (m_height * 0.5f));
+	}
```
Comment (karnkaul) [inline src/player.cpp]: Use `std::clamp` instead

File: src/target.hpp [hunk 27]
```diff
@@ -0,0 +1,42 @@
...
+	/// @param player_width: 
+	/// @param player_height
+	/// @return Type: Bool - Weather the target object has been hit or not.
+	/////////////////////////////////////////////////////////////
+	bool collision_hit(const sf::Vector2f& player_position, const uint16_t player_width, const uint16_t player_height);
+
+private:
+	float m_radius{};
+
+	std::default_random_engine m_engine;
```
Comment (karnkaul) [inline src/target.hpp]: This isn't seeded with anything, it will generate the same sequence every time. Use `m_engine{std::random_device{}()}` to seed it.

File: src/main.cpp (hunk 23 above)
Comment (JodyLS27) [reply src/main.cpp:189]: Good Idea, Was going to do this in my next build version with a full class rework.
Happy to do it now though.

File: src/main.cpp (hunk 24 above)
Comment (JodyLS27) [reply src/main.cpp:37]: On it, Will Adjust it and move forward, Thanks Karn :)

File: src/player.cpp (hunk 26 above)
Comment (JodyLS27) [reply src/player.cpp]: Yea cool, Better Method will switch to that. Didnt know it was there :)

File: src/target.hpp (hunk 27 above)
Comment (JodyLS27) [reply src/target.hpp]: Will Adjust this now.

Comment (JodyLS27) [reply src/target.hpp]: ✅ Addressed: Moved out of the constructor and Initialized in the header.

File: src/player.cpp (hunk 25 above)
Comment (JodyLS27) [reply src/player.cpp:96]: I don't quite understand what you mean by earlier?
The `set_direction()` function is the first thing to run after the inputs are mapped and set on that frame.

File: src/player.cpp (hunk 26 above)
Comment (JodyLS27) [reply src/player.cpp]: ✅ Addressed: Cleaned up and Optimized

**Comment by JodyLS27** [comment]
Going to close this PR, Too many issues with the code that needs cleanup and Organisation.
I will make a new PR in the Future.
